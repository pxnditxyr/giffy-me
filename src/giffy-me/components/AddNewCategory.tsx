import { FormEvent, useState } from 'react';
import { useForm } from '../hooks/useForm';

interface AddNewCategoryProps {
    currentCategories: Array<string>;
    onAddNewCategory: ( category : string ) => void;
}

export const AddNewCategory = ( { currentCategories, onAddNewCategory } : AddNewCategoryProps ) => {

    const [ isSame, setIsSame ] = useState<Boolean>( false );
    
    const { category, handleInputChange, handleReset } = useForm({
        category: ''
    });

    const handleSubmit = ( event : FormEvent<HTMLFormElement> ) => {
        event.preventDefault();
        
        const newCategory = category.trim();

        if ( newCategory.length < 2 ) return;

        if ( currentCategories.map( currentCategory => currentCategory.toLowerCase() ).includes( newCategory.toLowerCase() ) ) {
            setIsSame( true );
            return;
        }

        if ( isSame ) setIsSame( false );

        onAddNewCategory( newCategory );
        handleReset();
    };

    return (
        <form
            onSubmit={ handleSubmit }
            >
            <label> What are you thinking </label>
            <input
                type="text"
                name="category"
                value={ category }
                onChange={ handleInputChange }
                />
            {
                isSame 
                    && 
                ( <p> You already have this category </p> )
            }
            <button type="button"> Search </button>
        </form>
    );
};
