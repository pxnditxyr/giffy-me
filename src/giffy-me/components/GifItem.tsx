
interface GifItemProps {
    // id: string;
    title: string;
    url: string;
}

export const GifItem = ( { title, url } : GifItemProps ) => {

    return (
        <article
            className="flex flex-col justify-center items-center"
        >
            <img
                src={ url }
                alt={ title }
                className="w-full h-auto object-cover rounded-lg"
                />
            <p
                className="text-center text-xl font-bold text-gray-300 py-4"
            > { title } </p>
        </article>
    );
};
