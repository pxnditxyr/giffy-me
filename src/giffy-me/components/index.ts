export * from './AddNewCategory';
export * from './GifGrid';
export * from './GifItem';
