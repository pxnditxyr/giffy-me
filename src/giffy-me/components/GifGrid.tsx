import { GifItem } from './GifItem';

import { useFetchGifs } from '../hooks/useFetchGifs';

interface GifGridProps {
    category: string;
}

export const GifGrid = ( { category } : GifGridProps ) => {

    const { gifs, isLoading } = useFetchGifs( category );

    return (
        <div
            className="w-full flex flex-col justify-center items-center px-5"
        >
            <h3 
                className="text-center text-4xl font-bold text-amber-400 py-4 md:py-20"
            > { category } </h3>
            {
                isLoading
                    ? ( <p> Loading... </p> )
                    : (
                        <div
                            className="grid gap-4 grid-cols-1 md:grid-cols-2 lg:grid-cols-3 px-4"
                        >
                            {
                                gifs.map( ( gif ) => (
                                    <GifItem 
                                        key={ gif.id }
                                        { ...gif }
                                    />
                                ))
                            }
                        </div>
                    )
            }
        </div>
    );
};
