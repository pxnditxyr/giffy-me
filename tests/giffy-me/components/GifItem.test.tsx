import { render, screen } from '@testing-library/react';

import { GifItem } from '../../../src/giffy-me/components/GifItem';

describe( 'Tests in GifItem Component', () => {
    const title = 'Naruto';
    const url = 'https://www.imdb.com/title/tt0988824/';

    test( 'should render GifItem component', () => {
        const { container } = render( <GifItem title={ title } url={ url } /> );
        expect( container ).toMatchSnapshot();
    });

    test( 'should show image with url and alt provided', () =>{
        render( <GifItem title={ title } url={ url } /> );
        // screen.debug();
        const { src, alt } : HTMLImageElement = screen.getByRole( 'img' );
        expect( src ).toBe( url );
        expect( alt ).toBe( title );
    })

    test( 'should show title provided', () => {
        render( <GifItem title={ title } url={ url } /> );
        expect( screen.getByText( title ) ).toBeTruthy();
    });

});
