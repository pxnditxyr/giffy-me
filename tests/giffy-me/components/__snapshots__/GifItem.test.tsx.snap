// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in GifItem Component should render GifItem component 1`] = `
<div>
  <article
    class="flex flex-col justify-center items-center"
  >
    <img
      alt="Naruto"
      class="w-full h-auto object-cover rounded-lg"
      src="https://www.imdb.com/title/tt0988824/"
    />
    <p
      class="text-center text-xl font-bold text-gray-300 py-4"
    >
       
      Naruto
       
    </p>
  </article>
</div>
`;
